lib/semantics/check-select-stmt.cc

```diff
@@ -0,0 +1,112 @@
+//==check-select-stmt.cc - Checker for select-case, select-rank
+// TODO:select-type
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===--------------------------------------------------------------------------------------===
+
+#include "check-select-stmt.h"
+#include "tools.h"
+#include "../include/flang/ISO_Fortran_binding.h"
+#include "../parser/message.h"
+#include "../parser/tools.h"
+#include <list>
+#include <optional>
+#include <tuple>
+#include <variant>
+#include <vector>
+
+namespace Fortran::semantics {
+
+void SelectStmtChecker::Leave(
+    const parser::SelectRankConstruct &selectRankConstruct) {
+  const auto &selectRankStmt{
+      std::get<parser::Statement<parser::SelectRankStmt>>(
+          selectRankConstruct.t)};
+  const auto &selectRankStmtSel{
+      std::get<parser::Selector>(selectRankStmt.statement.t)};
+  // R1149 select-rank-stmt checks
+  if (std::holds_alternative<parser::Expr>(selectRankStmtSel.u)) {
+    // TODO:
+    // 1.Checks if selector is an expression ex.Rank(a+b)
+    //
+  } else if (const auto *variable{
+                 std::get_if<parser::Variable>(&selectRankStmtSel.u)}) {
+    if (const Symbol * entity{GetLastName(*variable).symbol}) {
+      if (!IsAssumedRankArray(*entity)) {  // C1150
+        context_.Say(variable->GetSource(),
+            "Selector is not an Assumed-rank array variable"_err_en_US);
+      }
+    }
+  }
+
+  // R1150 select-rank-case-stmt checks
+  auto &rankCaseList{std::get<std::list<parser::SelectRankConstruct::RankCase>>(
+      selectRankConstruct.t)};
+  bool defaultRankFound{false};
+  bool starRankFound{false};
+  std::vector<std::optional<std::int64_t>> matches;
```

Comment:
Why are you using a `vector` to represent a set?  Use `std::set` or `std::bitset`.